scheduler/utilization_based_host_allocator.go

```diff
@@ -54,15 +75,60 @@ func UtilizationBasedHostAllocator(ctx context.Context, hostAllocatorData HostAl
 	return newHostsNeeded, nil
 }
 
+// groupByTaskGroup takes a list of hosts and tasks and returns them grouped by task group
+func groupByTaskGroup(runningHosts []host.Host, taskQueue []model.TaskQueueItem) map[string]taskGroupData {
+	tgs := map[string]taskGroupData{}
+	for _, h := range runningHosts {
+		groupString := ""
+		if h.RunningTask != "" && h.RunningTaskGroup != "" {
+			groupString = makeTaskGroupString(h.RunningTaskGroup, h.LastBuildVariant, h.LastProject, h.LastVersion)
+		}
+		if data, exists := tgs[groupString]; exists {
+			data.hosts = append(data.hosts, h)
+			tgs[groupString] = data
+		} else {
+			tgs[groupString] = taskGroupData{
+				hosts: []host.Host{h},
+				tasks: []model.TaskQueueItem{},
+			}
+		}
+	}
+	for _, t := range taskQueue {
+		groupString := ""
+		if t.Group != "" {
+			groupString = makeTaskGroupString(t.Group, t.BuildVariant, t.Project, t.Version)
+
+		}
+		if data, exists := tgs[groupString]; exists {
+			data.tasks = append(data.tasks, t)
+			tgs[groupString] = data
+		} else {
+			tgs[groupString] = taskGroupData{
+				hosts: []host.Host{},
+				tasks: []model.TaskQueueItem{t},
+			}
+		}
+	}
+
+	return tgs
+}
+
+func makeTaskGroupString(group, bv, project, version string) string {
+	return fmt.Sprintf("%s_%s_%s_%s", group, bv, project, version)
+}
```

Comment:
shoiuldn't this just be a method on task?

Reply:
This is used for both tasks and hosts